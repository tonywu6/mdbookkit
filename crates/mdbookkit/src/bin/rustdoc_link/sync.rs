use std::{
    sync::{Arc, RwLock},
    task::Poll,
    time::Duration,
};

use anyhow::{bail, Result};
use tokio::{
    sync::{mpsc, Notify},
    task::JoinHandle,
    time,
};

pub struct EventSampling {
    pub buffer: Duration,
    pub timeout: Duration,
}

impl EventSampling {
    pub fn using<T>(self, rx: mpsc::Receiver<Poll<T>>) -> EventSampler<T>
    where
        T: Clone + Send + Sync + 'static,
    {
        EventSampler::new(rx, self)
    }
}

#[derive(Debug, Clone)]
pub struct EventSampler<T> {
    state: Arc<RwLock<State<T>>>,
    event: Arc<Notify>,
}

#[derive(Debug, Clone)]
enum State<T> {
    Pending,
    Ready(T),
    Timeout,
}

impl<T: Clone + Send + Sync + 'static> EventSampler<T> {
    fn new(
        mut rx: mpsc::Receiver<Poll<T>>,
        EventSampling { buffer, timeout }: EventSampling,
    ) -> Self {
        let state = Arc::new(RwLock::new(State::Pending));
        let event = Arc::new(Notify::new());

        tokio::spawn({
            let state = state.clone();
            let event = event.clone();
            async move {
                let mut abort: Option<JoinHandle<()>> = None;
                while let Some(value) = time::timeout(timeout, rx.recv()).await.transpose() {
                    if let Some(abort) = abort.take() {
                        abort.abort();
                    }
                    match value {
                        Ok(Poll::Ready(value)) => {
                            let event = event.clone();
                            let state = state.clone();
                            abort = Some(tokio::spawn(async move {
                                time::sleep(buffer).await;
                                *state.write().unwrap() = State::Ready(value);
                                event.notify_waiters();
                            }));
                        }
                        Ok(Poll::Pending) => {
                            *state.write().unwrap() = State::Pending;
                            event.notify_waiters();
                        }
                        Err(_) => {
                            *state.write().unwrap() = State::Timeout;
                            event.notify_waiters();
                        }
                    }
                }
            }
        });

        Self { event, state }
    }

    pub async fn wait(&self) -> Result<T> {
        loop {
            {
                match self.state.read().unwrap().clone() {
                    State::Pending => {}
                    State::Ready(value) => return Ok(value),
                    State::Timeout => bail!("timed out waiting for ready event"),
                }
            }
            self.event.notified().await;
        }
    }
}
