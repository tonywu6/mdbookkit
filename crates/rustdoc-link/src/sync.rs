use std::{
    sync::{Arc, RwLock},
    task::Poll,
    time::Duration,
};

use tokio::{
    sync::{mpsc, Notify},
    task::JoinHandle,
    time::sleep,
};

#[derive(Debug, Clone)]
pub struct Subsample<T> {
    state: Arc<RwLock<Poll<T>>>,
    event: Arc<Notify>,
}

impl<T: Clone + Send + Sync + 'static> Subsample<T> {
    pub fn new(mut rx: mpsc::Receiver<Poll<T>>, wait: Duration) -> Self {
        let state = Arc::new(RwLock::new(Poll::Pending));
        let event = Arc::new(Notify::new());

        tokio::spawn({
            let state = state.clone();
            let event = event.clone();
            async move {
                let mut abort: Option<JoinHandle<()>> = None;
                while let Some(value) = rx.recv().await {
                    if let Some(abort) = abort.take() {
                        abort.abort();
                    }
                    if value.is_ready() {
                        let event = event.clone();
                        let state = state.clone();
                        abort = Some(tokio::spawn(async move {
                            sleep(wait).await;
                            *state.write().unwrap() = value;
                            event.notify_waiters();
                        }));
                    } else {
                        *state.write().unwrap() = value;
                    }
                }
            }
        });

        Self { event, state }
    }

    pub async fn wait(&self) -> T {
        loop {
            {
                if let Poll::Ready(value) = self.state.read().unwrap().clone() {
                    return value;
                }
            }
            self.event.notified().await;
        }
    }
}
