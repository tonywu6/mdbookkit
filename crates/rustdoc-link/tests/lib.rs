use std::sync::Arc;

use anyhow::{bail, Result};

use console::set_colors_enabled_stderr;
use log::LevelFilter;
use mdbook_rustdoc_link::{
    env::{Config, Environment},
    logger::ConsoleLogger,
    Client, Pages, Resolver,
};
use similar::{ChangeTag, TextDiff};
use tap::Pipe;
use tokio::task::JoinSet;
use util_testing::{portable_snapshots, test_document, TestDocument};

async fn snapshot(client: Arc<Client>, TestDocument { source, name }: TestDocument) -> Result<()> {
    let stream = client.env().markdown(source).into_offset_iter();
    let mut page = Pages::one(source, stream)?;
    client.resolve(&mut page).await?;
    let output = page.get(&client.env().emit_config())?.to_string();

    portable_snapshots!().test(|| insta::assert_snapshot!(name.clone(), output))?;
    assert_no_whitespace_change(source, &output)?;

    let report = page
        .reporter()
        .level(LevelFilter::Trace)
        .paths(|_| name.clone())
        .report()
        .unwrap();
    portable_snapshots!().test(|| insta::assert_snapshot!(format!("{name}.stderr"), report))?;

    Ok(())
}

fn assert_no_whitespace_change(source: &str, output: &str) -> Result<()> {
    let changed_lines = TextDiff::from_words(source, output)
        .iter_all_changes()
        .filter_map(|change| {
            if matches!(change.tag(), ChangeTag::Equal) {
                return None;
            }
            if change.value().contains('\n') {
                Some(change.value())
            } else {
                None
            }
        })
        .collect::<Vec<_>>();

    if !changed_lines.is_empty() {
        bail!("unexpected whitespace change: {changed_lines:?}")
    } else {
        Ok(())
    }
}

#[tokio::test]
async fn test_snapshots() -> Result<()> {
    let client = setup()?;

    let tests = [
        test_document!("../../../docs/src/rustdoc-link/supported-syntax.md"),
        test_document!("../../../docs/src/rustdoc-link.md"),
        test_document!("tests/ra-known-quirks.md"),
    ];

    let errors = tests
        .map(|test| snapshot(client.clone(), test))
        .into_iter()
        .collect::<JoinSet<_>>()
        .join_all()
        .await
        .into_iter()
        .filter_map(Result::err)
        .collect::<Vec<_>>();

    if !errors.is_empty() {
        let errors = errors
            .iter()
            .map(|e| format!("{e:?}"))
            .collect::<Vec<_>>()
            .join("\n");
        panic!("{errors}")
    }

    client.drop().await?;

    Ok(())
}

fn setup() -> Result<Arc<Client>> {
    set_colors_enabled_stderr(false);
    ConsoleLogger::install();
    Config {
        rust_analyzer: Some("cargo run --release --package util-rust-analyzer -- analyzer".into()),
        ..Default::default()
    }
    .pipe(Environment::new)?
    .pipe(Client::new)
    .pipe(Arc::new)
    .pipe(Ok)
}
