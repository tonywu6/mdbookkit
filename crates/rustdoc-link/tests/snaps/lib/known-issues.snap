---
source: crates/rustdoc-link/tests/lib.rs
expression: output
---
# Known issues

<details class="toc" open>
  <summary>Sections</summary>

- [Performance](#performance)
- [Incorrect links](#incorrect-links)
  - [Macros](#macros)
  - [Trait items](#trait-items)
- [Unresolved items](#unresolved-items)
  - [Associated items on primitive types](#associated-items-on-primitive-types)
  - [Conditionally compiled items](#conditionally-compiled-items)
- [Sites other than docs.rs](#sites-other-than-docsrs)
- [Wrong line numbers in diagnostics](#wrong-line-numbers-in-diagnostics)

</details>

## Performance

`mdbook-rustdoc-link` itself doesn't need much processing power, but it invokes
rust-analyzer, which does a full scan of your workspace. The larger your codebase is,
the longer `mdbook` will have to wait for the preprocessor. This is the source of the
majority of the run time.

There is an [experimental caching feature](caching.md#enabling-caching), which persists
query results after runs and reuses them when possible, avoiding spawning rust-analyzer
when your edit doesn't involve item links.

## Incorrect links

In limited circumstances, rust-analyzer generates links that are incorrect or
inaccessible. Unfortunately, for such items, you will still have to link by hand.

> [!NOTE]
>
> The following observations are based on rust-analyzer
> <ra-version>(version)</ra-version>.

### Macros

Macros exported with [`#[macro_export]`][macro_export] are always exported at crate
root, and are documented as such by rustdoc, but rust-analyzer currently generates links
to the modules they are defined in. For example:

- [~~`panic!`~~](https://doc.rust-lang.org/stable/std/macros/macro.panic.html "panic!"), and many other `std` macros
  - The correct link is
    [https://doc.rust-lang.org/stable/std~~/macros~~/macro.panic.html][panic]
- [~~`serde_json::json!`~~](https://docs.rs/serde_json/1.0.140/serde_json/macros/macro.json.html "serde_json::json!")
  - The correct link is
    [https://docs.rs/serde_json/1.0.140/serde_json~~/macros~~/macro.json.html][serde_json::json]

Attribute macros generate links that use `macro.<macro_name>.html`, but rustdoc actually
generates `attr.<macro_name>.html`. For example:

- [~~`tokio::main!`~~](https://docs.rs/tokio-macros/2.5.0/tokio_macros/macro.main.html "tokio::main!")
  - The correct link is
    [https://docs.rs/tokio-macros/2.5.0/tokio_macros/~~macro~~attr.main.html][tokio::main]

### Trait items

Rust allows methods to have the same name if they are from different traits, and types
can implement the same trait multiple times if the trait is generic. All such methods
will appear on the same page for the type.

rustdoc will number the generated URL fragments so that they remain unique within the
HTML document. rust-analyzer does not yet have the ability to do so.

For example, these are the same links:

- [`<std::net::IpAddr as From<std::net::Ipv4Addr>>::from`](https://doc.rust-lang.org/stable/core/net/ip_addr/enum.IpAddr.html#method.from "<std::net::IpAddr as From<std::net::Ipv4Addr>>::from")
- [`<std::net::IpAddr as From<std::net::Ipv6Addr>>::from`](https://doc.rust-lang.org/stable/core/net/ip_addr/enum.IpAddr.html#method.from "<std::net::IpAddr as From<std::net::Ipv6Addr>>::from")

The correct link for the `From<Ipv6Addr>` implementation is actually
<a href="https://doc.rust-lang.org/stable/core/net/enum.IpAddr.html#method.from-1">
https://doc.rust-lang.org/stable/core/net/enum.IpAddr.html#method.from<strong>-1</strong>
</a>

## Unresolved items

### Associated items on primitive types

> [!NOTE]
>
> The following observations are based on rust-analyzer
> <ra-version>(version)</ra-version>.

Links to associated methods and items on primitive types are currently not resolved by
rust-analyzer. For example:

- \[`str::parse`]
- \[`f64::MIN_POSITIVE`]

### Conditionally compiled items

You may wish to link to items that are only conditionally compiled.

For example, [clap](https://docs.rs/clap/4.5.32/clap/index.html "clap") is known for providing guide-level documentation through docs.rs. To
link to the [tutorial for its Derive API][clap::_derive], theorectically you could write
\[`clap::_derive::_tutorial`]. However, this is gated behind the `clap/unstable-doc`
feature.

rust-analyzer supports configuring features to activate, but `mdbook-rustdoc-link` does
not yet support configuring rust-analyzer.

## Sites other than docs.rs

Currently, items from crates other than `std` always generate links that point to
<https://docs.rs>. `mdbook-rustdoc-link` does not yet support configuring alternative
hosting sites for crates (such as `wasm-bindgen` which hosts API docs under
`https://rustwasm.github.io/wasm-bindgen/api/`).

## Wrong line numbers in diagnostics

When the preprocessor fails to resolve some items, it emits warnings that look like:

<figure>

![warning emitted that has the wrong line numbers](media/error-reporting-wrong-line.png)

</figure>

You may notice that the line numbers are sometimes incorrect for your source file. This
could happen in files that use the `{{#include}}` directive, for example.

This is an unfortunate limitation with mdBook's preprocessor architecture. Preprocessors
are run sequentially, with the next preprocessor receiving Markdown source rendered by
the previous one. If preprocessors running before `mdbook-rustdoc-link` modify Markdown
source in such ways that shift lines around, then the line numbers will look incorrect.

Unless mdBook somehow gains [source map][sourcemap] support, this problem is unlikely to
ever be solved.

<!-- prettier-ignore-start -->

[macro_export]: https://doc.rust-lang.org/stable/reference/macros-by-example.html#path-based-scope
[panic]: https://doc.rust-lang.org/stable/std/macro.panic.html
[serde_json::json]: https://docs.rs/serde_json/1.0.140/serde_json/macro.json.html
[tokio::main]: https://docs.rs/tokio-macros/2.5.0/tokio_macros/attr.main.html
[IpV6Addr]: https://doc.rust-lang.org/stable/core/net/enum.IpAddr.html#method.from-1
[clap::_derive]: https://docs.rs/clap/4.5.32/clap/_derive/_tutorial/index.html
[sourcemap]: https://developer.mozilla.org/en-US/docs/Glossary/Source_map

<!-- prettier-ignore-end -->
